src/effect/delay.rs

```diff
@@ -0,0 +1,90 @@
+use std::sync::mpsc::Sender;
+
+use crate::{
+    audio::midi,
+    audio_unit::{self, delay::Message, AudioUnit},
+    config::DelayConfig,
+    effect::Effect,
+    Result,
+};
+use cpal::StreamConfig;
+use wmidi::MidiMessage;
+
+pub struct Delay {
+    config: DelayConfig,
+    unit: audio_unit::Split,
+    message_senders: Vec<Sender<Message>>,
+}
+
+impl Delay {
+    pub fn new(config: DelayConfig, stream_config: &StreamConfig) -> Result<Self> {
+        let mut audio_units = vec![];
+        let mut message_senders = vec![];
+
+        for n in 1..=config.num {
+            let (delay_unit, messages) =
+                audio_unit::Delay::new(stream_config, config.delay_ms * n)?;
+            let gain_unit = audio_unit::Gain::new(config.level.powi(n as i32)).boxed();
+            let pipeline = audio_unit::Pipeline::new(vec![delay_unit.boxed(), gain_unit])?.boxed();
+
+            audio_units.insert(audio_units.len(), pipeline);
+            message_senders.insert(message_senders.len(), messages);
+        }
+
+        let transparent = audio_unit::Transparent::new().boxed();
+        audio_units.insert(audio_units.len(), transparent);
```

Comment:
I'm not totally clear on conceptually why any given effect would hardcode the fact that it passes through the transparent signal

Reply:
I thought that a delay effect's output would also include the original (non-delayed) input, so it uses transparent to accomplish that. Maybe that is incorrect, though?

Reply:
I think it typically would. It just seems like an interesting thing (especially in terms of the mechanics of combining-primitive-units) to think about both generally and in terms of specific effects (a la how a wet/dry knob is a thing)